discovery/ovhcloud/ovhcloud.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovhcloud
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"inet.af/netaddr"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/regexp"
+	"github.com/ovh/go-ovh/ovh"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// metaLabelPrefix is the meta prefix used for all meta labels.
+// in this discovery.
+const (
+	metaLabelPrefix = model.MetaLabelPrefix + "ovhcloud_"
+)
+
+var (
+	ovhCloudApplicationKeyTest    = "TDPKJdwZwAQPwKX2"
+	ovhCloudApplicationSecretTest = config.Secret("9ufkBmLaTQ9nz5yMUlg79taH0GNnzDjk")
+	ovhCloudConsumerKeyTest       = "5mBuy6SUQcRw2ZUxg0cG68BoDKpED4KY"
+)
+
+type refresher interface {
+	refresh(context.Context) ([]*targetgroup.Group, error)
+}
+
+var DefaultSDConfig = SDConfig{
+	Endpoint:        "ovh-eu",
+	RefreshInterval: model.Duration(60 * time.Second),
+}
+
+// SDConfig sd config
+type SDConfig struct {
+	Endpoint          string         `yaml:"endpoint"`
+	ApplicationKey    string         `yaml:"application_key"`
+	ApplicationSecret config.Secret  `yaml:"application_secret"`
+	ConsumerKey       string         `yaml:"consumer_key"`
+	RefreshInterval   model.Duration `yaml:"refresh_interval"`
+	Service           string         `yaml:"service"`
+}
+
+// IPs struct to store IPV4 and IPV6
+type IPs struct {
+	IPV4 string `json:"ipv4"`
+	IPV6 string `json:"ipv6"`
+}
+
+// Name get name
+func (c SDConfig) Name() string {
+	return "ovhcloud"
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	if c.Endpoint == "" {
+		return errors.New("endpoint can not be empty")
+	}
+	if c.ApplicationKey == "" {
+		return errors.New("application key can not be empty")
+	}
+	if c.ApplicationSecret == "" {
+		return errors.New("application secret can not be empty")
+	}
+	if c.ConsumerKey == "" {
+		return errors.New("consumer key can not be empty")
+	}
+	switch c.Service {
```

Comment:
I feel like a if would be more readable here.

Reply:
We went with a switch for consistency since it appears that's how it is done for the others SDs, see:
https://github.com/prometheus/prometheus/blob/main/discovery/hetzner/hetzner.go#L95
https://github.com/prometheus/prometheus/blob/main/discovery/scaleway/scaleway.go#L61

I don't mind the modification, so it's up to you.